Persephone/Controllers/QueueViewController.swift

```diff
@@ -0,0 +1,183 @@
+//
+//  QueueController.swift
+//  Persephone
+//
+//  Created by Daniel Barber on 2019/2/01.
+//  Copyright © 2019 Dan Barber. All rights reserved.
+//
+
+import Cocoa
+import mpdclient
+
+class QueueViewController: NSViewController, NSOutlineViewDataSource, NSOutlineViewDelegate {
+  var queue: [MPDClient.Song] = []
+  var queuePos: Int32 = -1
+
+  var queueIcon: NSImage? = nil
+
+  let systemFontRegular = NSFont.systemFont(ofSize: 13, weight: .regular)
+  let systemFontBold = NSFont.systemFont(ofSize: 13, weight: .bold)
+
+  struct SongItem {
+    var song: MPDClient.Song
+    var queuePos: Int
+  }
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    queueView.columnAutoresizingStyle = .uniformColumnAutoresizingStyle
+
+    NotificationCenter.default.addObserver(
+      self,
+      selector: #selector(stateChanged(_:)),
+      name: MPDClient.stateChanged,
+      object: AppDelegate.mpdClient
+    )
+
+    NotificationCenter.default.addObserver(
+      self,
+      selector: #selector(queueChanged(_:)),
+      name: MPDClient.queueChanged,
+      object: AppDelegate.mpdClient
+    )
+
+    NotificationCenter.default.addObserver(
+      self,
+      selector: #selector(queuePosChanged(_:)),
+      name: MPDClient.queuePosChanged,
+      object: AppDelegate.mpdClient
+    )
+  }
+
+  @objc func stateChanged(_ notification: Notification) {
+    guard let state = notification.userInfo?[MPDClient.stateKey] as? mpd_state
+      else { return }
+
+    setQueueIcon(state)
+  }
+
+  @objc func queueChanged(_ notification: Notification) {
+    guard let queue = notification.userInfo?[MPDClient.queueKey] as? [MPDClient.Song]
+      else { return }
+
+    self.queue = queue
+
+    queueView.reloadData()
+  }
+
+  @objc func queuePosChanged(_ notification: Notification) {
+    guard let queuePos = notification.userInfo?[MPDClient.queuePosKey] as? Int32
+      else { return }
+
+    let oldSongRowPos = Int(self.queuePos + 1)
+    let newSongRowPos = Int(queuePos + 1)
+    self.queuePos = queuePos
```

Comment:
Two things: Do you have to explicitly cast to `Int` here?

Do you think it would be more understandable (and less prone to future bugs) to have a differently-named local variable for `queuePos` so you don't have to keep differentiating with `self`?

Reply:
The queue position returned by `libmpdclient` is an `Int32` so I'm storing it as such here in the controller. I only need it as an `Int` when accessing rows of the table view. I add one because the first row is a heading.

Reply:
I think the dynamic cast above on line 70 could be `as? Int`, and the `queuePos` instance property also changed to `Int` (an `Int` will always fit at least an `Int32`, so the cast should succeed).

Reply:
Interesting. I assumed an `Int` would be 16-bit by default, but it looks like it's actually 64-bit (or 32-bit on 32-bit platforms).

Reply:
@sharplet I tried this but it didn't work until I explicitly cast to `Int` somewhere. Ended up doing it in the status model.